pkg/app/test_command.go

```diff
@@ -6,130 +6,136 @@ import (
 // titleFilterTitle is the title of test which should be executed, if empty it will execute all tests
 // ctx holds the command flags. If directory scanning is enabled with --dir it is
 // not supported to filter tests, therefore testFilterTitle is an empty string
 func TestCommand(testPath string, testFilterTitle string, ctx AddCommandContext) error {
-	if ctx.Verbose == true {
+	if ctx.Verbose {
 		log.SetOutput(os.Stdout)
 	}
 
+	out = output.NewCliOutput(!ctx.NoColor)
+
 	if testPath == "" {
 		testPath = CommanderFile
 	}
 
-	var results <-chan runtime.TestResult
+	var result runtime.Result
 	var err error
-	if ctx.Dir {
+	switch {
+	case ctx.Dir:
 		if testFilterTitle != "" {
 			return fmt.Errorf("Test may not be filtered when --dir is enabled")
 		}
 		fmt.Println("Starting test against directory: " + testPath + "...")
 		fmt.Println("")
-		results, err = testDir(testPath)
-	} else {
+		result, err = testDir(testPath)
+	default:
 		fmt.Println("Starting test file " + testPath + "...")
 		fmt.Println("")
-		results, err = testFile(testPath, testFilterTitle)
+		result, err = testFile(testPath, "", testFilterTitle)
 	}
 
 	if err != nil {
 		return fmt.Errorf(err.Error())
 	}
 
-	out := output.NewCliOutput(!ctx.NoColor)
-	if !out.Start(results) {
+	if !out.PrintSummary(result) && !ctx.Verbose {
 		return fmt.Errorf("Test suite failed, use --verbose for more detailed output")
 	}
 
 	return nil
 }
 
-func testDir(directory string) (<-chan runtime.TestResult, error) {
+func testDir(directory string) (runtime.Result, error) {
+	result := runtime.Result{}
+
+	// TODO: validate entire dir before running tests: issue #129
 	files, err := ioutil.ReadDir(directory)
 	if err != nil {
-		return nil, fmt.Errorf(err.Error())
+		return result, fmt.Errorf(err.Error())
 	}
 
-	results := make(chan runtime.TestResult)
-	var wg sync.WaitGroup
-
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		for _, f := range files {
-			// Skip reading dirs for now. Should we also check valid file types?
-			if f.IsDir() {
-				continue
-			}
-
-			fileResults, err := testFile(path.Join(directory, f.Name()), "")
-			if err != nil {
-				panic(fmt.Sprintf("%s: %s", f.Name(), err))
-			}
-
-			for r := range fileResults {
-				r.FileName = f.Name()
-				results <- r
-			}
+	for _, f := range files {
+		if f.IsDir() {
+			continue // skip dirs: TODO add support for walking
+		}
+
+		path := path.Join(directory, f.Name())
+		newResult, err := testFile(path, f.Name(), "")
+		if err != nil {
+			return result, err
 		}
-	}()
 
-	go func(ch chan runtime.TestResult) {
-		wg.Wait()
-		close(results)
-	}(results)
+		result = convergeResults(result, newResult)
+	}
 
-	return results, nil
+	return result, nil
 }
 
-func testFile(filePath string, title string) (<-chan runtime.TestResult, error) {
-	content, err := readFile(filePath)
+func convergeResults(result runtime.Result, new runtime.Result) runtime.Result {
```

Comment:
This is something which could be done inside the `runtime` package in the future. For now it is okay that it the composition and aggregation is inside the `test_command.go`.

Reply:
opened #141